### PR TITLE
Use non-generic TaskCompletionSource where possible in System.Threading.Channels

### PR DESCRIPTION
**Pull request by danielmarbach**

### CODE REVIEW

**Comment by danielmarbach** [comment]
@Dotnet-GitSync-Bot area-System.Threading.Channels

File: src/libraries/System.Threading.Channels/src/System/Threading/Channels/BoundedChannel.cs [hunk 1]
```diff
@@ -16,7 +16,7 @@ internal sealed class BoundedChannel<T> : Channel<T>, IDebugEnumerable<T>
         /// <summary>The mode used when the channel hits its bound.</summary>
         private readonly BoundedChannelFullMode _mode;
         /// <summary>Task signaled when the channel has completed.</summary>
-        private readonly TaskCompletionSource<VoidResult> _completion;
+        private readonly TaskCompletionSource _completion;
```
Comment (stephentoub) [inline src/libraries/System.Threading.Channels/src/System/Threading/Channels/BoundedChannel.cs:19]: These changes won't work in the netstandard build of the library, where there is no non-generic TCS.  You might try creating a TCS in that build, e.g. add a file that's only compiled into that build and that provides a polyfill:
```C#
internal sealed class TaskCompletionSource : TaskCompletionSource<VoidResult> { ... }
```

Comment (danielmarbach) [reply src/libraries/System.Threading.Channels/src/System/Threading/Channels/BoundedChannel.cs:19]: Yeah haven't thought about the netstandard part. Do you think it is worth doing so or would you as a maintainer prefer to stay with the current `TaskCompletionSource<VoidResult>`? I have no problems just closing this or do what you suggested to add a polyfill

Comment (stephentoub) [reply src/libraries/System.Threading.Channels/src/System/Threading/Channels/BoundedChannel.cs:19]: It's not worth complicating the code with ifdefs are partial files or anything like that.  But if such a polyfill addresses it, that'd be file.

Comment (danielmarbach) [reply src/libraries/System.Threading.Channels/src/System/Threading/Channels/BoundedChannel.cs:19]: Pushed. I hope I got it right. Sorry this is the first time I contribute here

Comment (stephentoub) [reply src/libraries/System.Threading.Channels/src/System/Threading/Channels/BoundedChannel.cs:19]: Thanks!

**Comment by danielmarbach** [comment]
@stephentoub ready for review

**Comment by danielmarbach** [comment]
It is missing the area labels and I don't know how to add those

File: src/libraries/System.Threading.Channels/src/System/Threading/Channels/TaskCompletionSource.cs [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        /// The <see cref="Tasks.Task"/> created by this instance and accessible through its <see cref="Task"/> property
+        /// will be instantiated using the specified <paramref name="creationOptions"/>.
+        /// </remarks>
+        /// <param name="creationOptions">The options to use when creating the underlying <see cref="Tasks.Task"/>.</param>
+        /// <exception cref="ArgumentOutOfRangeException">
+        /// The <paramref name="creationOptions"/> represent options invalid for use
+        /// with a <see cref="TaskCompletionSource"/>.
+        /// </exception>
+        public TaskCompletionSource(TaskCreationOptions creationOptions) :
+            base(null, creationOptions)
```
Comment (stephentoub) [inline src/libraries/System.Threading.Channels/src/System/Threading/Channels/TaskCompletionSource.cs:22]:
```suggestion
            base(creationOptions)
```

File: src/libraries/System.Threading.Channels/src/System/Threading/Channels/TaskCompletionSource.cs [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+        /// <remarks>
+        /// This operation will return false if the <see cref="Tasks.Task"/> is already in one of the three final states:
+        /// <see cref="TaskStatus.RanToCompletion"/>,
+        /// <see cref="TaskStatus.Faulted"/>, or
+        /// <see cref="TaskStatus.Canceled"/>.
+        /// </remarks>
+        public bool TrySetResult()
+        {
+            return TrySetResult(default);
+        }
```
Comment (stephentoub) [inline src/libraries/System.Threading.Channels/src/System/Threading/Channels/TaskCompletionSource.cs:39]:
```suggestion
        public bool TrySetResult() => TrySetResult(default);
```

File: src/libraries/System.Threading.Channels/src/System/Threading/Channels/TaskCompletionSource.cs (hunk 2 above)
Comment (stephentoub) [inline src/libraries/System.Threading.Channels/src/System/Threading/Channels/TaskCompletionSource.cs:22]:
```suggestion
        public TaskCompletionSource(TaskCreationOptions creationOptions) : base(creationOptions)
```

**Review by stephentoub — APPROVED** [review]
Other than my comments, LGTM. Thanks!

**Comment by stephentoub** [comment]
Thanks!

**Pull request merged**